polkadot/node/core/pvf/prepare-worker/src/lib.rs

```diff
@@ -324,3 +200,174 @@ fn runtime_construction_check(
+					Ok(output)
+				});
+			}
+			result
+		},
+		Arc::clone(&condvar),
+		WaitOutcome::Finished
+	).unwrap_or_else(|_| {
+		process::exit(libc::EXIT_FAILURE)
+	});
+
+	let outcome = thread::wait_for_threads(condvar);
+
+	match outcome {
+		WaitOutcome::Finished => {
+			 let result = prepare_thread.join().unwrap_or_else(|_| {
+				(
+					process::exit(libc::EXIT_FAILURE)
+				)
+			});
+            cfg_if::cfg_if! {
+				if #[cfg(target_os = "linux")] {
+					let (artifact_result, max_rss) = result;
+				} else {
+					let artifact_result = result;
+				}
+			}
+
+			// Stop the memory stats worker and get its observed memory stats.
+			#[cfg(any(target_os = "linux", feature = "jemalloc-allocator"))]
+				let memory_tracker_stats = get_memory_tracker_loop_stats(memory_tracker_thread, worker_pid).await;
+
+			let memory_stats = MemoryStats {
+				#[cfg(any(target_os = "linux", feature = "jemalloc-allocator"))]
+				memory_tracker_stats,
+				#[cfg(target_os = "linux")]
+				max_rss: extract_max_rss_stat(max_rss, worker_pid),
+			};
+
+			let response = Response { artifact_result, memory_stats };
+
+			let bytes = bincode::serialize(&response).unwrap_or_else(|_| {
+				process::exit(libc::EXIT_FAILURE)
+			});
+
+			pipe_write.write_all(bytes.as_slice()).unwrap_or_else(|_| {
+				process::exit(libc::EXIT_FAILURE)
+			});
+
+			process::exit(libc::EXIT_SUCCESS);
+
+		},
+		_ => {
+			process::exit(libc::EXIT_FAILURE)
+		}
+	}
+}
+
+async fn handle_parent_process(
+	mut pipe_read: os_pipe::PipeReader,
+	pipe_write: PipeWriter,
+	temp_artifact_dest: PathBuf,
+	preparation_timeout: Duration,
+	worker_pid: u32,
+) -> Result<PrepareStats, PrepareError> {
+	drop(pipe_write);
+
+	return match nix::sys::wait::wait() {
+		Ok(nix::sys::wait::WaitStatus::Exited(_, libc::EXIT_SUCCESS)) => {
+			let data_size = mem::size_of::<Response>();
+
+			let mut received_data = Vec::new();
+
+			pipe_read.read_to_end(&mut received_data).map_err(|_| {
+				PrepareError::Panic(format!("error reading pipe for worker id {}", worker_pid))
+			})?;
+
+			let result: Response = bincode::deserialize(received_data.as_slice()).map_err(|e| {
+				PrepareError::Panic(e.to_string())
+			})?;
+
+			drop(pipe_read);
+
+			match result.artifact_result {
+				Err(err) => Err(err),
+				Ok(artifact) => {
+					// Write the serialized artifact into a temp file.
+					//
+					// PVF host only keeps artifacts statuses in its memory,
+					// successfully compiled code gets stored on the disk (and
+					// consequently deserialized by execute-workers). The prepare worker
+					// is only required to send `Ok` to the pool to indicate the
+					// success.
+					gum::debug!(
+						target: LOG_TARGET,
+						%worker_pid,
+						"worker: writing artifact to {}",
+						temp_artifact_dest.display(),
+					);
+					if let Err(err) = tokio::fs::write(&temp_artifact_dest, &artifact).await {
+						return Err(PrepareError::Panic(format!("{:?}", err)))
```

Comment:
We're over-using `Panic` error types. I will take of it later, it involves some analysis of whether the error conditions are deterministic or not etc.

Reply:
I renamed `Panic` to `JobError` which we use to catch any unexpected error that happened during the job. After all, there's nothing special about a panic for us to treat it differently, it's just one possible unexpected error that could happen.